primitives/arithmetic/src/lib.rs

```diff
@@ -85,8 +86,328 @@ where
 	}
 }
 
+/// A collection-like object that is made of values of type `T` and can normalize its individual
+/// values around a centric point.
+pub trait Normalizable<T> {
+	/// Normalize self around `targeted_sum`.
+	///
+	/// Only returns `Ok` if the new sum of results is guaranteed to be equal to `targeted_sum`.
+	/// Else, returns an error explaining why it failed to do so.
+	fn normalize(&self, targeted_sum: T) -> Result<Vec<T>, &'static str>;
+}
+
+macro_rules! impl_normalize_for_numeric {
+	($numeric:ty) => {
+		impl Normalizable<$numeric> for Vec<$numeric> {
+			fn normalize(&self, targeted_sum: $numeric) -> Result<Vec<$numeric>, &'static str> {
+				normalize(self.as_ref(), targeted_sum)
+			}
+		}
+	};
+}
+
+impl_normalize_for_numeric!(u32);
+impl_normalize_for_numeric!(u64);
+impl_normalize_for_numeric!(u128);
+
+impl<P: PerThing> Normalizable<P> for Vec<P> {
+	fn normalize(&self, targeted_sum: P) -> Result<Vec<P>, &'static str> {
+		let inners = self.iter().map(|p| p.clone().deconstruct().into()).collect::<Vec<_>>();
+		let normalized = normalize(inners.as_ref(), targeted_sum.deconstruct().into())?;
+		Ok(normalized.into_iter().map(|i: UpperOf<P>| P::from_parts(i.saturated_into())).collect())
+	}
+}
+
+
+/// Normalize `input` so that the sum of all elements reaches `targeted_sum`.
+///
+/// This implementation is currently in a balanced position between being performant and accurate.
+///
+/// 1. We prefer storing original indices, and sorting the `input` only once. This will save the
+///    cost of sorting per round at the cost of a little bit of memory.
+/// 2. The granularity of increment/decrements is determined by the difference and the number of
+///    elements in `input` and their sum difference with `targeted_sum`. In short, the
+///    implementation is very likely to over-increment/decrement in case the `diff / input.len()` is
+///    rather small.
+///
+/// This function can return an error is if size of `input` doesn't fit inside `T`, or if
+/// `sum(input)` cannot fit inside `T`. Moreover, if any of the internal operations saturate, it
+/// will also return en `Err`.
+///
+/// Based on the above, the best use case of the function will be only to correct small rounding
+/// errors. If the difference of the elements is vert large, then the subtraction/addition of one
+/// element with `diff / input.len()` might easily saturate and results will be `Err`.
+pub fn normalize<T>(input: &[T], targeted_sum: T) -> Result<Vec<T>, &'static str>
```

Comment:
does this works with signed integer ?

Reply:
Nope.

Reply:
If I try to use `num_traits::Unsigned`, it would require some unrelated traits to be implemented for `Per_thing`, and then we can't have `Normalizable<Per_thing>`. I think the only other way is to use our own custom Unsigned trait.

Reply:
Limited now with `num_traits::Unsigned`.